vlasovsolver/cpu_moments.cpp

```diff
@@ -152,75 +164,84 @@ void calculateMoments_R(
-          }
-
-          const Real dx = cell->parameters[CellParams::DX];
-          const Real dy = cell->parameters[CellParams::DY];
-          const Real dz = cell->parameters[CellParams::DZ];
-
-          vmesh::VelocityBlockContainer<vmesh::LocalID>& blockContainer = cell->get_velocity_blocks(popID);
-          if (blockContainer.size() == 0) continue;
-          const Realf* data       = blockContainer.getData();
-          const Real* blockParams = blockContainer.getParameters();
-          const Real mass = getObjectWrapper().particleSpecies[popID].mass;
-          const Real charge = getObjectWrapper().particleSpecies[popID].charge;
-
-          #ifdef DEBUG_MOMENTS
-          bool ok = true;
-          if (data == NULL && blockContainer.size() > 0) ok = false;
-          if (blockParams == NULL && blockContainer.size() > 0) ok = false;
-          if (ok == false) {
-             stringstream ss;
-             ss << "ERROR in moment calculation in " << __FILE__ << ":" << __LINE__ << endl;
-             ss << "\t &data = " << data << "\t &blockParams = " << blockParams << endl;
-             ss << "\t size = " << blockContainer.size() << endl;
-             cerr << ss.str();
-             exit(1);
-          }
-          #endif
-
-          // Temporary array where the moments for this species are accumulated
-          Real array[4];
-          for (int i=0; i<4; ++i) array[i] = 0.0;
-
-          // Calculate species' contribution to first velocity moments
-          for (vmesh::LocalID blockLID=0; blockLID<blockContainer.size(); ++blockLID) {
-            blockVelocityFirstMoments(data+blockLID*WID3,
-                                      blockParams+blockLID*BlockParams::N_VELOCITY_BLOCK_PARAMS,
-                                      array);
-          } // for-loop over velocity blocks
-
-          // Store species' contribution to bulk velocity moments
-          Population & pop = cell->get_population(popID);
-          pop.RHO_R = array[0];
-          pop.V_R[0] = divideIfNonZero(array[1], array[0]);
-          pop.V_R[1] = divideIfNonZero(array[2], array[0]);
-          pop.V_R[2] = divideIfNonZero(array[3], array[0]);
-          
-          cell->parameters[CellParams::RHOM_R  ] += array[0]*mass;
-          cell->parameters[CellParams::VX_R] += array[1]*mass;
-          cell->parameters[CellParams::VY_R] += array[2]*mass;
-          cell->parameters[CellParams::VZ_R] += array[3]*mass;
-          cell->parameters[CellParams::RHOQ_R  ] += array[0]*charge;
-       } // for-loop over spatial cells
+       if (getObjectWrapper().particleSpecies[popID].propagateSpecies == true) {
+
+           #pragma omp parallel for
+           for (size_t c=0; c<cells.size(); ++c) {
+              SpatialCell* cell = mpiGrid[cells[c]];
+              
+	      if (cell->sysBoundaryFlag == sysboundarytype::DO_NOT_COMPUTE) {
+		 continue;
+	      }
+              // Clear old moments to zero value
+              if (popID == 0) {
```

Comment:
Wow, this could break if a test species is defined before. We need another mechanism here, to be safe, or we'll be scratching our heads deeply when this bites us.

Reply:
Or rather a non-propagated species. I saw now that the treatment of test vs non-test and propagated vs non-propagated species is not consistent between the _R and _V moment computations. Is this intentional? It's too late for me to wrap my head around this now.

Reply:
Good catch, that would indeed break. And I think the thing about _R and _V treating non-propagated species differently was a mistake. I've reworked this section and tried to consolidate the two phases.